Working Streamer

QuentinI opened the pull request:
Closes #<ISSUE_NUMBER>
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->

### This PR does not:
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

### Key places to review:
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1209939621908045
  - https://app.asana.com/0/0/1209980878478912
  - https://app.asana.com/0/0/1209980878478924
  - https://app.asana.com/0/0/1209980878478918

QuentinI committed:
Fix docker builds

QuentinI committed:
WIP

QuentinI committed:
Remove some of the workarounds

QuentinI commented on espresso/batch_buffer.go:
```diff
@@ -0,0 +1,79 @@
+type Batch interface {
+	Number() uint64
+	L1Origin() eth.BlockID
+	Hash() common.Hash
```
Artifact of a workaround, can be removed

shenkeyao replied:
I didn't remove this yet, because according to https://github.com/EspressoSystems/optimism-espresso-integration/pull/96#discussion_r2040253533, it seems that we'll need some refactoring after removing this.

QuentinI replied:
Oh, I meant Hash() function, it's not necessary for this interface.

QuentinI commented on espresso/environment/espresso_dev_node_test.go:
```diff
@@ -99,7 +100,7 @@ func runSimpleL1TransferAndVerifier(ctx context.Context, t *testing.T, system *e
 
 	// Send Transaction on L1, and wait for verification on the L2 Verifier
-	ctx, cancel := context.WithTimeout(ctx, 15*time.Second)
+	ctx, cancel := context.WithTimeout(ctx, 2*time.Minute)
```
Tests pass on my machine in about 30secs, but that's still too slow

QuentinI commented on espresso/environment/espresso_dev_node_test.go:
```diff
@@ -127,6 +128,7 @@ func runSimpleL1TransferAndVerifier(ctx context.Context, t *testing.T, system *e
 			t.Errorf("waiting for balance change returned with error:\nhave:\n\t\"%v\"\nwant:\t\n\"%v\"\n", have, want)
 		}
 
+		log.Println("endBalance:", endBalance, "startBalance:", startBalance)
```
A lot of extraneous logging here can be removed, used for debugging

shenkeyao replied:
Follow-up task: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209980878478912?focus=true

QuentinI commented on espresso/streamer.go:
```diff
@@ -22,28 +18,27 @@ type L1Client interface {
 }
 
-type EspressoStreamer struct {
+type EspressoStreamer[B Batch] struct {
```
Generic over batch type, so we can define a batch - which depends on types from op-node - in op-node, but everything else gets it's own package. Keeps op-node changes smaller.
Theoretically we could build on it and split it out into it's own repository later for future integrations to use

dailinsubjam replied:
Love this.

shenkeyao replied:
Follow-up: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209980878478920?focus=true

QuentinI commented on espresso/streamer.go:
```diff
@@ -58,49 +53,73 @@ type EspressoStreamer struct {
-	if err != nil {
-		return false, err
-	}
+	//	hotshotState, err := s.EspressoLightClient.LightClient.
```
Light client still doesn't work in tests, so for now I just set confirmedHotShotPos to 0 here. Could be either replaced by fixed light client or by manually tracking Espresso blocks we've seen that have corresponding SafeL2 as @Ayiga proposed

shenkeyao replied:
Follow-up: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209980878478924?focus=true

QuentinI commented on espresso/streamer.go:
```diff
@@ -110,48 +129,72 @@ func (s *EspressoStreamer) Update(ctx context.Context) error {
+				continue
+			}
+
+			// origin := (*batch).L1Origin()
```
Starts of an attempt to ensure sequencer proposed batch with finalized L1 state.
In general, this section of the code is unfinished, probably needs more checks for batch validity.

shenkeyao replied:
Follow-up: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209980878478927?focus=true

QuentinI commented on espresso/streamer.go:
```diff
@@ -188,21 +231,19 @@ func (s *EspressoStreamer) Start(ctx context.Context) error {
-func (s *EspressoStreamer) Next(ctx context.Context) *EspressoBatchI {
+func (s *EspressoStreamer[B]) Next(ctx context.Context) *B {
+	if s.BatchBuffer.Len() > 0 && (*s.BatchBuffer.Peek()).Number() != s.BatchPos {
+		log.Warn("Batch buffer is being weird", "batchPos", s.BatchPos, "batchBuffer", s.BatchBuffer)
```
Extraneous logging, nothing weird about it. Added when I was debugging with strictly sequential batch posting.

shenkeyao replied:
Removed in https://github.com/EspressoSystems/optimism-espresso-integration/pull/96/commits/99e32de72a7e322476c2183523d45f47cf0d014c.

QuentinI commented on op-batcher/batcher/driver.go:
```diff
@@ -124,6 +124,9 @@ type BatchSubmitter struct {
 	channelMgr      *channelManager
 	prevCurrentL1   eth.L1BlockRef // cached CurrentL1 from the last syncStatus
+
+	blocks     map[common.Hash]*types.Block
```
Can be removed.

shenkeyao replied:
Removed in https://github.com/EspressoSystems/optimism-espresso-integration/pull/96/commits/99e32de72a7e322476c2183523d45f47cf0d014c.

QuentinI commented on op-batcher/batcher/espresso.go:
```diff
@@ -45,42 +44,43 @@ func (l *BatchSubmitter) tryPublishBatchToEspresso(ctx context.Context, transact
 func (l *BatchSubmitter) queueBlockToEspresso(ctx context.Context, block *types.Block) error {
-	batch, _, err := derive.BlockToSingularBatch(l.RollupConfig, block)
+	l.blockMutex.Lock()
+	l.blocks[block.Hash()] = block
```
Can be removed

shenkeyao replied:
Removed in https://github.com/EspressoSystems/optimism-espresso-integration/pull/96/commits/99e32de72a7e322476c2183523d45f47cf0d014c.

QuentinI commented on op-node/node/node.go:
```diff
@@ -587,11 +587,11 @@ func (n *OpNode) Start(ctx context.Context) error {
 
 	if n.cfg.Rollup.CaffNodeConfig.IsCaffNode {
-		go func() {
-			if err := n.l2Driver.SyncDeriver.Derivation.EspressoStreamer().Start(ctx); err != nil {
```
Espresso streamer either needs to get Start back - and gain a Mutex protecting it - or we need to drive it manually in op-node, similar to what we do in batcher.

QuentinI replied:
Personally, I'm in favor of self-driving streamer, we could also switch it to websocket API to make it faster.

shenkeyao replied:
No strong preference. It looks like a self-driving streamer would require fewer changes but we need to be more careful about concurrency handling.

shenkeyao replied:
Follow-up: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209980878478930?focus=true

QuentinI commented on op-node/rollup/derive/espresso_batch.go:
```diff
@@ -43,158 +37,26 @@ func getFinalizedL1(header *espressoCommon.HeaderImpl) *espressoCommon.L1BlockIn
-	Batch  SingularBatch
+	Header        *types.Header
+	Batch         SingularBatch
+	L1InfoDeposit *types.Transaction
```
This set of fields looks enough for channel manager to post correct batches

QuentinI commented on op-node/rollup/derive/pipeline_espresso.go:
Can be removed, was playing around with op-node and accidentally commited.

shenkeyao replied:
Removed in https://github.com/EspressoSystems/optimism-espresso-integration/pull/96/commits/99e32de72a7e322476c2183523d45f47cf0d014c.

philippecamacho committed:
Document how to authenticate to docker using a PAT

shenkeyao commented on op-batcher/batcher/espresso.go:
```diff
@@ -158,25 +159,33 @@ func (l *BatchSubmitter) espressoBatchLoadingLoop(ctx context.Context, wg *sync.
+				// 	panic("Not our block")
+				// }
+
 				// This should happen ONLY if the batch is malformed. BatchToIncompleteBlock has to guarantee
```
Nit: Should `BatchToIncompleteBlock` be `ToBlock`?

shenkeyao replied:
Addressed in https://github.com/EspressoSystems/optimism-espresso-integration/pull/96/commits/558e0e9f923cce1b19bb9e5ce7621c52a90a55cf.

shenkeyao commented on espresso/environment/optitmism_espresso_test_helpers.go:
```diff
@@ -28,11 +28,7 @@ import (
-
-// deployed ESPRESSO_SEQUENCER_LIGHT_CLIENT_ADDRESS at 0x17435cce3d1b4fa2e5f8a08ed921d57c6762a180
-// deployed ESPRESSO_SEQUENCER_PLONK_VERIFIER_ADDRESS at 0xb4b46bdaa835f8e4b4d8e208b6559cd267851051
-const ESPRESSO_LIGHT_CLIENT_ADDRESS = "0x17435cce3d1b4fa2e5f8a08ed921d57c6762a180"
```
Nit: We can keep this `const` but use the correct address.

shenkeyao replied:
Addressed in https://github.com/EspressoSystems/optimism-espresso-integration/pull/96/commits/558e0e9f923cce1b19bb9e5ce7621c52a90a55cf.

shenkeyao commented on espresso/streamer.go:
```diff
@@ -58,49 +53,73 @@ type EspressoStreamer struct {
-	s.Log.Debug("Updated finalized hotshot state", "hotshotState", hotshotState)
-
-	targetHeight := min(hotshotState.BlockHeight, s.hotShotPos+100)
+	targetHeight := min(blockHeight, s.hotShotPos+100)
```
Note that with the issue and workaround mentioned at https://github.com/EspressoSystems/optimism-espresso-integration/pull/96/files#r2040255786, `hotShotPos` will be 100 after `Reset` is called, so `targetHeight` is at most 100.

shenkeyao replied:
Follow-up: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209980878478924?focus=true

shenkeyao commented on espresso/streamer.go:
Note for cleanup: We should revisit log levels in this PR before merging. No need to do this before addressing other comments.

shenkeyao replied:
Follow-up: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209980878478912?focus=true

shenkeyao commented on espresso/streamer.go:
```diff
@@ -110,48 +129,72 @@ func (s *EspressoStreamer) Update(ctx context.Context) error {
-		}
+		// var header espressoTypes.HeaderImpl
+		// err = json.Unmarshal(rawHeader, &header)
+		// if err != nil {
+		// 	return fmt.Errorf("could not unmarshal header from bytes")
+		// }
 
-		snapshot, err := s.EspressoLightClient.FetchMerkleRoot(s.hotShotPos, nil)
-		if err != nil {
-			return fmt.Errorf("failed to fetch Merkle root: %w", err)
-		}
+		// snapshot, err := s.EspressoLightClient.FetchMerkleRoot(s.hotShotPos, nil)
+		// if err != nil {
+		// 	return fmt.Errorf("failed to fetch Merkle root: %w", err)
+		// }
 
-		if snapshot.Height <= s.hotShotPos {
-			return fmt.Errorf("snapshot height is less than or equal to the requested height")
-		}
+		// if snapshot.Height <= s.hotShotPos {
+		// 	return fmt.Errorf("snapshot height is less than or equal to the requested height")
+		// }
```
Note for cleanup: Restore these commented-out lines before merging.

shenkeyao replied:
Follow-up: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209980878478927?focus=true

shenkeyao commented on espresso/streamer.go:
```diff
@@ -110,48 +129,72 @@ func (s *EspressoStreamer) Update(ctx context.Context) error {
 
 		// TODO Philippe initialize when creating the streamer
-		s.BatchBuffer.SetBatcherAddress(s.BatcherAddress)
+		//s.BatchBuffer.SetBatcherAddress(s.BatcherAddress)
```
Note for cleanup: Restore this.

shenkeyao replied:
`SetBatcherAddress` has been removed from the interface.

shenkeyao committed:
Remove unused types and file

shenkeyao committed:
Merge branch 'ag/l1finality' of github.com:EspressoSystems/optimism-espresso-integration into ag/l1finality

shenkeyao committed:
Address my cleanup comments

shenkeyao committed:
Merge branch 'celo-integration' into ag/l1finality

shenkeyao commented on op-batcher/batcher/espresso.go:
```diff
@@ -132,7 +132,7 @@ func (l *BatchSubmitter) espressoBatchLoadingLoop(ctx context.Context, wg *sync.
 		l.L1Client,
 		l.Espresso,
-		l.EspressoLightClient,
+		nil, // TODO (Keyao) BatchSubmitter doesn't have field EspressoLightClient.
```
Note this TODO. We need to either remove this field or fix the `nil` depending on whether to use the light client.

shenkeyao replied:
Follow-up: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209980878478924?focus=true

philippecamacho committed:
Update justfile to run the smoke tests.

philippecamacho committed:
just command to stop all the espresso containers.

philippecamacho committed:
Attempt to have the Espresso container work with Linux

Ayiga committed:
Attempt to fix some linux docker container parameters

QuentinI committed:
Cleanup

philippecamacho commented on espresso/environment/optitmism_espresso_test_helpers.go:
```diff
@@ -28,7 +28,9 @@ import (
-const ESPRESSO_DEV_NODE_DOCKER_IMAGE = "ghcr.io/espressosystems/espresso-sequencer/espresso-dev-node:release-goldendoodle"
+//const ESPRESSO_DEV_NODE_DOCKER_IMAGE = "ghcr.io/espressosystems/espresso-sequencer/espresso-dev-node:release-goldendoodle"
+
+const ESPRESSO_DEV_NODE_DOCKER_IMAGE = "ghcr.io/espressosystems/espresso-sequencer/espresso-dev-node:20250409-dev-node-pos-preview"
```
So we are using the most recent image now?

QuentinI committed:
Remove some extraneous logging

Ayiga committed:
Reverse a change to http virtual host lists

QuentinI committed:
Fix test

philippecamacho committed:
Add localhost to the list of HTTPVirtualHosts.

philippecamacho committed:
Remove redundant field BatcherAddress in CaffNodeConfig.

philippecamacho committed:
Display Espresso go library version when entering nix shell.

philippecamacho committed:
Update justfile to run all the integration tests and delete the Espresso containers.

shenkeyao committed:
Merge branch 'celo-integration' into ag/l1finality

Ayiga committed:
Add Espresso Block Conversion Test

In order to better validate that our stored representation of data
from the L2 does not lose information, or result in different
information when stored, it would be beneficial to have a test to
validate these concerns.

This commit adds a test to ensure that our EspressoBlock is able
to be constructed from a `Block` and rollup info  and then
converted back again with the passed rollup info.

QuentinI committed:
Bring Light Client back

QuentinI committed:
Merge branch 'celo-integration' into ag/l1finality
